models/migrations/v101.go

```diff
@@ -0,0 +1,74 @@
+// Copyright 2019 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package migrations
+
+import (
+	"crypto/md5"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"code.gitea.io/gitea/modules/setting"
+
+	"xorm.io/xorm"
+)
+
+func renameExistingUserAvatarName(x *xorm.Engine) error {
+	sess := x.NewSession()
+	defer sess.Close()
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	type User struct {
+		ID     int64 `xorm:"pk autoincr"`
+		Avatar string
+	}
+
+	users := make([]User, 0)
+	if err := sess.Find(&users); err != nil {
+		return err
+	}
+
+	deleteList := make([]string, 0, len(users))
+	for _, user := range users {
+		oldAvatar := user.Avatar
+		newAvatar := fmt.Sprintf("%x", md5.Sum([]byte(fmt.Sprintf("%d-%s", user.ID, user.Avatar))))
+
+		if _, err := os.Stat(filepath.Join(setting.AvatarUploadPath, oldAvatar)); err != nil {
+			continue
+		}
+
+		fr, err := os.Open(filepath.Join(setting.AvatarUploadPath, oldAvatar))
+		if err != nil {
+			return err
```

Comment:
Any OS errors below this point should try to commit previous changes into the database. Also, deletion of successfully migrated entries could be attempted.

Please format the error so the user will know what was the operation that failed ("open", "read", "write", "delete", etc.).

If possible, try to commit every _n_ changes (e.g. _n_ = 100) to avoid potential problems with a big change transaction. Also, please check if the new file name is the same as the old's, in case this procedure is retried and some avatars were already migrated.

Reply:
> Any OS errors below this point should try to commit previous changes into the database.

> If possible, try to commit every n changes (e.g. n = 100) to avoid potential problems with a big change transaction.

If I directly use `x.Update(...)`, I don't have to consider if the session has big change transaction and it will solve the problem, right ?

> Also, deletion of successfully migrated entries could be attempted.

Deletion of successfully migrated entries should not be attempted until the whole migration is completed. We still don't know if any user still has avatar similar with previously migrated ones.

> Please format the error so the user will know what was the operation that failed ("open", "read", "write", "delete", etc.).

Yeah, we should totally do it.

> Also, please check if the new file name is the same as the old's, in case this procedure is retried and some avatars were already migrated.

As we are adding a prefix with old avatar and then hashing it, it will never be same as the previous avatar. If we follow the new naming system, I think there's no way to know if it's already migrated.
Besides, if we migrate already migrated avatars, will it cause any problem ?

Reply:
> 
> 
> > Any OS errors below this point should try to commit previous changes into the database.
> 
> > If possible, try to commit every n changes (e.g. n = 100) to avoid potential problems with a big change transaction.
> 
> If I directly use `x.Update(...)`, I don't have to consider if the session has big change transaction and it will solve the problem, right ?

You are right in theory, and I was initially going to propose that, but we've seen a couple of cases where many consecutive transactions could cause TCP port depletion problems (e.g. #8540). That's why I think going for a safer path, with less connections, is better. But I don't feel too comfortable with using one big transaction for all the changes either, that's why processing in batches sounds better. It's a technique that is already being used in other migration steps.

> 
> > Also, deletion of successfully migrated entries could be attempted.
> 
> Deletion of successfully migrated entries should not be attempted until the whole migration is completed. We still don't know if any user still has avatar similar with previously migrated ones.

You are right about this. I have no problem with some small unconnected files remaining in case of a migration fails half-way.

> 
> > Also, please check if the new file name is the same as the old's, in case this procedure is retried and some avatars were already migrated.
> 
> As we are adding a prefix with old avatar and then hashing it, it will never be same as the previous avatar. If we follow the new naming system, I think there's no way to know if it's already migrated.

The name of the file is now normalized by hashing the content (the inner MD5) with the user ID (outer MD5), so it will always render the same hash. **But** then I realize that you're not reading the file but trusting the previous hash. Perhaps it's worth re-hashing the file to make sure the MD5 matches the content?

> Besides, if we migrate already migrated avatars, will it cause any problem ?

I'm not sure, but it seems to me that Open() --> Create() ---> Copy() on the same path should cause some trouble. We'd need to test it with Linux, Windows, network file systems, small files, big files, etc. I think it's just easier to check if we are not trying to copy the file onto itself.

Reply:
> The name of the file is now normalized by hashing the content (the inner MD5) with the user ID (outer MD5), so it will always render the same hash. But then I realize that you're not reading the file but trusting the previous hash. Perhaps it's worth re-hashing the file to make sure the MD5 matches the content?

I tried it but didn't worked. At first I thought I am doing something wrong but then after many trial and errors, realized the `md5 hash` is generated just after uploading the image and that hash value is used as avatar name but the image is saved to file storage after some preprocessing. That's why the file never generates the same hash value.
https://github.com/go-gitea/gitea/blob/1ae25c086e9e085cf44b71720211bd836300b373/models/user.go#L510-L511

https://github.com/go-gitea/gitea/blob/1ae25c086e9e085cf44b71720211bd836300b373/models/user.go#L542-L544

> > As we are adding a prefix with old avatar and then hashing it, it will never be same as the previous avatar. If we follow the new naming system, I think there's no way to know if it's already migrated.
> 
> The name of the file is now normalized by hashing the content (the inner MD5) with the user ID (outer MD5), so it will always render the same hash. But then I realize that you're not reading the file but trusting the previous hash. Perhaps it's worth re-hashing the file to make sure the MD5 matches the content?
>
> > Besides, if we migrate already migrated avatars, will it cause any problem ?

So, I think we have no option but migrate all avatars.

Reply:
OK, I see your points.

Reply:
> Any OS errors below this point should try to commit previous changes into the database. 

> Please format the error so the user will know what was the operation that failed ("open", "read", "write", "delete", etc.).

> If possible, try to commit every n changes (e.g. n = 100) to avoid potential problems with a big change transaction

I did those changes. Please have a look.